Standalone Soda Core guide (#410)
astronomer/airflow-guides: guides/soda-data-quality.md

---
title: "Soda Core and Airflow"
description: "Using Soda Core to implement data quality checks in Airflow DAGs"
date: 2022-08-02T00:00:00.000Z
slug: "soda-data-quality"
heroImagePath: null
tags: ["Soda", "Data Quality"]
---

[Soda Core](https://www.soda.io/core) is an open source framework for data quality checks using the SodaCL interface to run checks defined in a YAML file.

Soda Core lets you:

> - Define checks as YAML configuration with many predefined metrics available.

Author: No, you are correct, metric is the property of the dataset that the check is run on. Thanks for catching that 😅

- Provide any SQL query within the YAML file and check against its returned value, if no predefined metrics fit your use case.
- Easily set up a large amount of data quality checks.
- Integrate data quality checks with commonly used data engineering tools such as Airflow, Apache Spark, PostgreSQL, Snowflake [and more](https://www.soda.io/integrations).

In this guide, we will cover key features of Soda Core and how to use it with Airflow.

> **Note**: For a general overview on how to approach data quality, and of different tools available to run data quality checks using Airflow, see the 'Data Quality and Airflow' guide.

## Assumed Knowledge

To get the most out of this guide, you should have knowledge of:

- How to design a data quality approach. See Data Quality and Airflow.
- Airflow operators. See [Operators 101](https://www.astronomer.io/guides/what-is-an-operator/).

> - Relational Databases. See [Relational database on Wikipedia](https://en.wikipedia.org/wiki/Relational_database).

Author: True! I wasn't sure what our policy is with linking to resources on general concepts from companies, that is why I went with wikipedia, but yours is better.

  - invalid_percent(WEBSITE_COL) < 5:
      valid regex: astronomer\.io
  # The average of 3 columns for values of category_1 is between 10 and 100
  - my_average_total_for_category_1 between 10 and 100:
      my_average_total_for_category_1 query: |
        SELECT AVG(MY_COL_1 + MY_COL_2 + MY_COL_3)
        FROM MY_TABLE_1
        WHERE MY_CATEGORY = 'category_1'
```

Three more unique features of Soda Core are freshness checks, schema checks and reference checks:

```YAML
checks for MY_TABLE_1:
  # MY_DATE has no values older than 10 days
  - freshness(MY_DATE) <= 10d
  # The schema has to have the MY_KEY column
  - schema:
      fail:
        when required column missing: [MY_KEY]
  # all names listed in MY_TABLE_1's MY_NAMES column have to also exists
  # in the MY_CUSTOMER_NAMES column in MY_TABLE_2
  - values in (MY_NAMES) must exist in MY_TABLE_2 (MY_CUSTOMER_NAMES)
```

## Requirements

To use Soda Core, the Soda Core package for the database backend needs to be installed and configured. The [Soda documentation provides a list of supported databases](https://docs.soda.io/soda-core/configuration.html) and how to configure them.

> When using Soda Core with Airflow, the Soda Core package needs to be installed in your Airflow environment. Additionally, two YAML files need to be made available to the Airflow environment: `configuration.yml`, which contains the connection information for the database backend, and `checks.yml`, which contains the data quality checks.

Reviewer: ```suggestion
When using Soda Core with Airflow, the Soda Core package needs to be installed in your Airflow environment. Additionally, two YAML files need to be made available to the Airflow environment: `configuration.yml`, which contains the connection information for the database backend, and `checks.yml`, which contains the data quality checks.

```

> Astro CLI users can simply put `soda-core-<database>` in `requirements.txt` and provide the YAML files in `/include`.

Reviewer: ```suggestion
If you use the Astro CLI, you can add `soda-core-<database>` to `requirements.txt` and your YAML files to `/include` in your Astro project.
```

## Example: Run Soda Core checks from Airflow

> This example shows one possible way to set up data quality checks on a Snowflake database using Soda Core from within Airflow.

Reviewer: ```suggestion
This example shows one possible way to run data quality checks on a Snowflake database using Soda Core from within Airflow.
```

- Create a checks file containing the data quality checks: `checks.yml`.
- Install the `soda-core-snowflake` package in your Airflow environment.
- Run the checks from a DAG by running the `soda scan` command using the `BashOperator`.

### Step 1: Create the configuration file

First we need to create a configuration file to connect to Snowflake. The easiest way to create the file is to use the template in the Soda documentation as shown below.

```YAML
# the first line names the datasource "MY_DATASOURCE"
data_source MY_DATASOURCE:
  type: snowflake
  connection:
    # provide your snowflake username and password in double quotes
    username: "MY_USERNAME"
    password: "MY_PASSWORD"
    # provide the account in the format xy12345.eu-central-1
    account: my_account
    database: MY_DATABASE
    warehouse: MY_WAREHOUSE
    # if your connection times out you may need to adjust the timeout value
    connection_timeout: 300
    role: MY_ROLE
    client_session_keep_alive:
    session_parameters:
      QUERY_TAG: soda-queries
      QUOTED_IDENTIFIERS_IGNORE_CASE: false
  schema: MY_SCHEMA
```

> Save the YAML instructions above in a filed called `configuration.yml` and make it available to your Airflow environment. Astro CLI users can place the file into the `/include` directory.

Reviewer: ```suggestion
Save the YAML instructions above in a filed called `configuration.yml` and make it available to your Airflow environment. If you use the Astro CLI, you can do this by placing the file into the `/include` directory of your Astro project.
```

### Step 2: Create the checks file

> You can define your data quality checks using the [many checks available for Soda CL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a predefined metric or check that works for your use case, you can create a user-defined check using SQL as shown below.

Reviewer: ```suggestion
You can define your data quality checks using the [many preset checks available for Soda CL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a preset metric or check that works for your use case, you can create a custom check using SQL as shown below.
```
Still confused about metric vs. check here

Author: Yeah, that is on me, sorry! I misread something and and thought things like missing_count or duplicate_count were called metrics and other checks were called checks. But as I said above metrics are the property of the data that is being checked. I'll change it in the text. 🙂

```YAML
checks for example_table:
  # check that MY_EMAIL_COL contains only email addresses according to the
  # format name@domain.extension
  - invalid_count(MY_EMAIL_COL) = 0:
      valid format: email
  # check that all entries in MY_DATE_COL are unique
  - duplicate_count(MY_DATE_COL) = 0
  # check that MY_TEXT_COL has no missing values
  - missing_count(MY_TEXT_COL) = 0
  # check that MY_TEXT_COL has at least 10 distinct values using SQL
  - distinct_vals >= 10:
      distinct_vals query: |
        SELECT DISTINCT(MY_TEXT_COL) FROM example_table
  # check that MY_NUM_COL has a minimum between 90 and 110
  - min(MY_NUM_COL) between 90 and 110
  # check that example table has at least 1000 rows
  - row_count >= 1000
  # check that the sum of MY_COL_2 is bigger than the sum of MY_COL_1
  - sum_difference > 0:
      sum_difference query: |
        SELECT SUM(MY_COL_2) - SUM(MY_COL_1) FROM example_table
  # checks that all entries in MY_COL_3 are part of a set or possible values
  - invalid_count(MY_COL_3) = 0:
      valid values: ['val1', 'val2', 'val3', 'val4']
```

> Save the YAML instructions above in a filed called `checks.yml` and make it available to your Airflow environment. Astro CLI users can place the file into the `/include` directory.

Reviewer: ```suggestion
Save the YAML instructions above in a filed called `checks.yml` and make it available to your Airflow environment. If you use the Astro CLI, you can place this file in your `/include` directory.
```

### Step 3: Install the Soda Core package

> The `soda-core-snowflake` package needs to be installed in your Airflow environment. Astro CLI users can add the following line to the `requirements.txt` file:

Reviewer: ```suggestion
The `soda-core-snowflake` package needs to be installed in your Airflow environment. If you use the Astro CLI, you can do this by adding the following line to the `requirements.txt` file:
```

```text
soda-core-snowflake
```

### Step 4: Run soda scan using the BashOperator

In a DAG, Soda Core checks are executed by using the `BashOperator` to run the `soda scan` command. The DAG below shows how to reference the configuration and checks YAML files in the command.

```python
from airflow import DAG
from datetime import datetime

from airflow.operators.bash import BashOperator

SODA_PATH="<filepath>" # can be specified as an env variable

with DAG(
    dag_id="soda_example_dag",
    schedule_interval='@daily',
    start_date=datetime(2022,8,1),
    catchup=False
) as dag:

    soda_test = BashOperator(
        task_id="soda_test",
        bash_command=f"soda scan -d MY_DATASOURCE -c \
            {SODA_PATH}/configuration.yml {SODA_PATH}/checks.yml"
    )
```

> The logs from the Soda Core checks can be found in the Airflow task logs. The logs will list all checks that ran along with their result.

Reviewer: ```suggestion
Because Soda Core runs through the BashOperator, you can run your checks in any part of your DAG and trigger tasks with your results like you would with other Airflow operators.

The logs from the Soda Core checks can be found in the Airflow task logs. The logs list all checks that ran and their results.
```

> Below an example of the logs in the case of 3 checks passing:

Reviewer: ```suggestion
The following is an example of how your logs might look if all 3 checks pass:
```

```text
[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO - Scan summary:
[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO - 3/3 checks PASSED:
[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO -     MY_TABLE in MY_DATASOURCE
[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO -       duplicate_count(MY_ID_COLUMN) = 0 [PASSED]
[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO -       missing_count(MY_ID_COLUMN) = 0 [PASSED]
[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO -       min(MY_NUM_COL) between 0 and 10 [PASSED]
[2022-08-04, 13:07:22 UTC] {subprocess.py:92} INFO - All is good. No failures. No warnings. No errors.
```

> In the case of a check failure, the logs will show which check failed and the `check_value` that caused the failure.

Reviewer: ```suggestion
In the case of a check failure, the logs show which check failed and the `check_value` that caused the failure:
```

[2022-08-04, 13:23:59 UTC] {subprocess.py:92} INFO -     MY_TABLE in MY_DATASOURCE
[2022-08-04, 13:23:59 UTC] {subprocess.py:92} INFO -       duplicate_count(MY_ID_COLUMN) = 0 [PASSED]
[2022-08-04, 13:23:59 UTC] {subprocess.py:92} INFO -       missing_count(MY_ID_COLUMN) = 0 [PASSED]
[2022-08-04, 13:23:59 UTC] {subprocess.py:92} INFO - 1/3 checks FAILED:
[2022-08-04, 13:23:59 UTC] {subprocess.py:92} INFO -     MY_TABLE in MY_DATASOURCE
[2022-08-04, 13:23:59 UTC] {subprocess.py:92} INFO -       max(MY_NUM_COL) between 10 and 20 [FAILED]
[2022-08-04, 13:23:59 UTC] {subprocess.py:92} INFO -         check_value: 3
[2022-08-04, 13:23:59 UTC] {subprocess.py:92} INFO - Oops! 1 failures. 0 warnings. 0 errors. 2 pass.
[2022-08-04, 13:24:00 UTC] {subprocess.py:96} INFO - Command exited with return code 2
[2022-08-04, 13:24:00 UTC] {taskinstance.py:1909} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/airflow/operators/bash.py", line 194, in execute
    raise AirflowException(
airflow.exceptions.AirflowException: Bash command failed. The command returned a non-zero exit code 2.
```

> Of course it is possible to incorporate data quality checks using Soda Core in different locations within your DAG, and to use the results of those checks to control the flow of data within your pipeline.

Reviewer: ```suggestion
```
If you want to keep this sentence, I think it's better to add it before the DAG code above. It feels a bit like a cliffhanger here.

Author: 🧗

---

[Soda Core](https://www.soda.io/core) is an open source framework for data quality checks using the SodaCL interface to run checks defined in a YAML file.

Soda Core lets you:

- Define checks as YAML configuration with many predefined metrics available.
- Provide any SQL query within the YAML file and check against its returned value, if no predefined metrics fit your use case.
- Easily set up a large amount of data quality checks.
- Integrate data quality checks with commonly used data engineering tools such as Airflow, Apache Spark, PostgreSQL, Snowflake [and more](https://www.soda.io/integrations).

In this guide, we will cover key features of Soda Core and how to use it with Airflow.

> **Note**: For a general overview on how to approach data quality, and of different tools available to run data quality checks using Airflow, see the 'Data Quality and Airflow' guide.

## Assumed Knowledge

To get the most out of this guide, users should have knowledge of:

- How to design a data quality approach
- Airflow operators
- Relational Databases
- Basic familiarity with writing YAML configurations

The following resources are recommended:

- Data Quality and Airflow

> - [Operators 101](Operators 101)

Reviewer 2: Update to the link to the guide here

Author: I could have sworn I did correct that... weird. Sorry! 😅

Soda Core offers the ability to run checks on different properties of your dataset against a numerically defined threshold:

```YAML
checks for MY_TABLE_1:
  # MY_NUM_COL_1 has a minimum of above or equal 0
  - min(MY_NUM_COL_1) >= 0
  # MY_TEXT_COL has less than 10% missing values
  - missing_percent(MY_TEXT_COL) < 10
checks for MY_TABLE_2:
  # MY_NUM_COL_2 has an average between 100 and 1000
  - avg(MY_NUM_COL_2) is between 100 and 1000
  # MY_ID_COL has no duplicates
  - duplicate_count(MY_ID_COL) = 0
```

It is possible to add optional check configurations like naming a check or configure warning versus failing conditions:

```YAML
checks for MY_TABLE_1:
  # fail the check when MY_TABLE_1 has less than 10 or more than a million rows
  # warn if there are less than 100 (but 10 or more) rows
  - row_count:
      warn: when < 100
      fail:
        when < 10
        when > 1000000
      name: Wrong number of rows!
```

Data can be filtered according to validity criteria using the following methods:

> - List of valid values.

Reviewer 2: I think we don't need the periods at the end of these bullet points since they aren't full sentences

  - invalid_percent(WEBSITE_COL) < 5:
      valid regex: astronomer\.io
  # The average of 3 columns for values of category_1 is between 10 and 100
  - my_average_total_for_category_1 between 10 and 100:
      my_average_total_for_category_1 query: |
        SELECT AVG(MY_COL_1 + MY_COL_2 + MY_COL_3)
        FROM MY_TABLE_1
        WHERE MY_CATEGORY = 'category_1'
```

Three more unique features of Soda Core are freshness checks, schema checks and reference checks:

```YAML
checks for MY_TABLE_1:
  # MY_DATE has no values older than 10 days
  - freshness(MY_DATE) <= 10d
  # The schema has to have the MY_KEY column
  - schema:
      fail:
        when required column missing: [MY_KEY]
  # all names listed in MY_TABLE_1's MY_NAMES column have to also exists
  # in the MY_CUSTOMER_NAMES column in MY_TABLE_2
  - values in (MY_NAMES) must exist in MY_TABLE_2 (MY_CUSTOMER_NAMES)
```

## Requirements

To use Soda Core, the Soda Core package for the database backend needs to be installed and configured. The [Soda documentation provides a list of supported databases](https://docs.soda.io/soda-core/configuration.html) and how to configure them.

> When using Soda Core with Airflow the Soda Core package needs to be installed in your Airflow environment. Additionally, both YAML files, `configuration.yml`, which contains the connection information for the database backend and `checks.yml`, which contains the data quality checks will need to be made available to the Airflow environment.

Reviewer 2: ```suggestion
When using Soda Core with Airflow, the Soda Core package needs to be installed in your Airflow environment. Additionally, two YAML files need to be made available to the Airflow environment: `configuration.yml`, which contains the connection information for the database backend, and `checks.yml`, which contains the data quality checks.
```

- Create a checks file containing the data quality checks: `checks.yml`.
- Install the `soda-core-snowflake` package in your Airflow environment.
- Run the checks from a DAG by running the `soda scan` command using the `BashOperator`.

### Step 1: Create the configuration file

First we need to create a configuration file to connect to Snowflake. The easiest way to create the file is to use the template in the Soda documentation as shown below.

```YAML
# the first line names the datasource "MY_DATASOURCE"
data_source MY_DATASOURCE:
  type: snowflake
  connection:
    # provide your snowflake username and password in double quotes
    username: "MY_USERNAME"
    password: "MY_PASSWORD"
    # provide the account in the format xy12345.eu-central-1
    account: my_account
    database: MY_DATABASE
    warehouse: MY_WAREHOUSE
    # if your connection times out you may need to adjust the timeout value
    connection_timeout: 300
    role: MY_ROLE
    client_session_keep_alive:
    session_parameters:
      QUERY_TAG: soda-queries
      QUOTED_IDENTIFIERS_IGNORE_CASE: false
  schema: MY_SCHEMA
```

> 

Reviewer 2: Let's add a sentence here that says something like "Paste the text above into a file called `configuration.yml` and place it in your `include/` directory"

```YAML
checks for example_table:
  # check that MY_EMAIL_COL contains only email addresses according to the
  # format name@domain.extension
  - invalid_count(MY_EMAIL_COL) = 0:
      valid format: email
  # check that all entries in MY_DATE_COL are unique
  - duplicate_count(MY_DATE_COL) = 0
  # check that MY_TEXT_COL has no missing values
  - missing_count(MY_TEXT_COL) = 0
  # check that MY_TEXT_COL has at least 10 distinct values using SQL
  - distinct_vals >= 10:
      distinct_vals query: |
        SELECT DISTINCT(MY_TEXT_COL) FROM example_table
  # check that MY_NUM_COL has a minimum between 90 and 110
  - min(MY_NUM_COL) between 90 and 110
  # check that example table has at least 1000 rows
  - row_count >= 1000
  # check that the sum of MY_COL_2 is bigger than the sum of MY_COL_1
  - sum_difference > 0:
      sum_difference query: |
        SELECT SUM(MY_COL_2) - SUM(MY_COL_1) FROM example_table
  # checks that all entries in MY_COL_3 are part of a set or possible values
  - invalid_count(MY_COL_3) = 0:
      valid values: ['val1', 'val2', 'val3', 'val4']
```

> > **Note**: You need to make both YAML files available to your Airflow environment. Users of the Astro CLI can simply put them into the `/include` folder.

Reviewer 2: Same comment here as above - instead of having a note about this, let's add an explicit sentence in both sections saying to put the text in a file called xyz and put it somewhere where it will be available to Airflow e.g. include/

---
title: "Soda Core and Airflow"

> description: "Using Soda Core to implement data quality checks in Airflow DAGs"

Author: ```suggestion
description: "Using Soda Core to implement data quality checks in Airflow DAGs."
```

> date: 2022-08-02T00:00:00.000Z

Author: ```suggestion
date: 2022-08-11T00:00:00.000Z
```

slug: "soda-data-quality"
heroImagePath: null

> tags: ["Soda", "Data Quality"]

Author: ```suggestion
tags: ["Soda", "Data quality"]
```

---

[Soda Core](https://www.soda.io/core) is an open source framework for checking data quality. It uses the Soda Checks Language (SodaCL) to run checks defined in a YAML file.

Soda Core lets you:

- Define checks as YAML configuration, including many preset checks.
- Provide a SQL query within the YAML file and check against a returned value if no preset checks fit your use case.
- Integrate data quality checks with commonly used data engineering tools such as Airflow, Apache Spark, PostgreSQL, Snowflake [and more](https://www.soda.io/integrations).

In this guide, we will cover key features of Soda Core and how to use it with Airflow.

> > **Note**: For a general overview on how to approach data quality, and of different tools available to run data quality checks using Airflow, see the 'Data Quality and Airflow' guide.

Author: ```suggestion
> **Note**: For a general overview on how to approach data quality, and of different tools available to run data quality checks using Airflow, see the '[Data quality and Airflow](https://www.astronomer.io/guides/data-quality)' guide.
```

## Assumed knowledge

To get the most out of this guide, you should have knowledge of:

> - How to design a data quality approach. See Data Quality and Airflow.

Author: ```suggestion
- How to design a data quality approach. See [Data Quality and Airflow](https://www.astronomer.io/guides/data-quality).
```

  - missing_count(IP_ADDRESS_COL) < 10:
      valid format: ip address
  # WEBSITE_COL has less than 5% entries that don't contain "astronomer.io"
  - invalid_percent(WEBSITE_COL) < 5:
      valid regex: astronomer\.io
  # The average of 3 columns for values of category_1 is between 10 and 100
  - my_average_total_for_category_1 between 10 and 100:
      my_average_total_for_category_1 query: |
        SELECT AVG(MY_COL_1 + MY_COL_2 + MY_COL_3)
        FROM MY_TABLE_1
        WHERE MY_CATEGORY = 'category_1'
```

Three more unique features of Soda Core are:

- [Freshness checks](https://docs.soda.io/soda-cl/freshness.html): Set limits to the age of the youngest row in the table.
- [Schema checks](https://docs.soda.io/soda-cl/schema.html): Run checks on the existence of columns and validate data types.
- [Reference checks](https://docs.soda.io/soda-cl/reference.html): Ensure parity in between columns in different datasets in the same data source.

```YAML
checks for MY_TABLE_1:
  # MY_DATE's youngest row is younger 10 days
  - freshness(MY_DATE) < 10d
  # The schema has to have the MY_KEY column
  - schema:
      fail:
        when required column missing: [MY_KEY]

>   # all names listed in MY_TABLE_1's MY_NAMES column have to also exists

Author: ```suggestion
  # all names listed in MY_TABLE_1's MY_NAMES column have to also exist
```

When using Soda Core with Airflow, the Soda Core package needs to be installed in your Airflow environment. Additionally, two YAML files need to be made available to the Airflow environment: `configuration.yml`, which contains the connection information for the database backend, and `checks.yml`, which contains the data quality checks.

If you use the Astro CLI, you can add `soda-core-<database>` to `requirements.txt` and your YAML files to `/include` in your Astro project.

## Example: Run Soda Core checks from Airflow

This example shows one possible way to run data quality checks on a Snowflake database using Soda Core from within Airflow.

The setup can be divided into the following steps:

- Create a configuration file that connects to the Snowflake database: `configuration.yml`.
- Create a checks file containing the data quality checks: `checks.yml`.
- Install the `soda-core-snowflake` package in your Airflow environment.
- Run the checks from a DAG by running the `soda scan` command using the `BashOperator`.

> 

Author: ```suggestion
```

### Step 1: Create the configuration file

First we need to create a configuration file to connect to Snowflake. The easiest way to create the file is to use the template in the Soda documentation as shown below.

```YAML
# the first line names the datasource "MY_DATASOURCE"
data_source MY_DATASOURCE:
  type: snowflake
  connection:
    # provide your snowflake username and password in double quotes
    username: "MY_USERNAME"
    password: "MY_PASSWORD"
    # provide the account in the format xy12345.eu-central-1
    account: my_account
    database: MY_DATABASE
    warehouse: MY_WAREHOUSE
    # if your connection times out you may need to adjust the timeout value
    connection_timeout: 300
    role: MY_ROLE
    client_session_keep_alive:
    session_parameters:
      QUERY_TAG: soda-queries
      QUOTED_IDENTIFIERS_IGNORE_CASE: false
  schema: MY_SCHEMA
```

> Save the YAML instructions above in a filed called `configuration.yml` and make it available to your Airflow environment. If you use the Astro CLI, you can do this by placing the file into the `/include` directory of your Astro project.

Author: ```suggestion
Save the YAML instructions above in a file called `configuration.yml` and make it available to your Airflow environment. If you use the Astro CLI, you can do this by placing the file into the `/include` directory of your Astro project.
```

### Step 2: Create the checks file

> You can define your data quality checks using the [many preset checks available for Soda CL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a preset check that works for your use case, you can create a custom one using SQL as shown below.

Author: ```suggestion
You can define your data quality checks using the [many preset checks available for SodaCL](https://docs.soda.io/soda-cl/soda-cl-overview.html). If you cannot find a preset check that works for your use case, you can create a custom one using SQL as shown below.
```

```text
soda-core-snowflake
```

### Step 4: Run soda scan using the BashOperator

In a DAG, Soda Core checks are executed by using the `BashOperator` to run the `soda scan` command. The DAG below shows how to reference the configuration and checks YAML files in the command.

```python
from airflow import DAG
from datetime import datetime

from airflow.operators.bash import BashOperator

SODA_PATH="<filepath>" # can be specified as an env variable

with DAG(
    dag_id="soda_example_dag",
    schedule_interval='@daily',
    start_date=datetime(2022,8,1),
    catchup=False
) as dag:

    soda_test = BashOperator(
        task_id="soda_test",
        bash_command=f"soda scan -d MY_DATASOURCE -c \
            {SODA_PATH}/configuration.yml {SODA_PATH}/checks.yml"
    )
```

Because Soda Core runs through the `BashOperator`, you can run your checks in any part of your DAG and trigger tasks with your results like you would with other Airflow operators.

The logs from the Soda Core checks can be found in the Airflow task logs. The logs list all checks that ran and their results.

> The following is an example of how your logs might look if all 3 checks pass:

Author: ```suggestion
This is an example of what your logs might look like if all 3 checks pass:
```
